mlrun/utils/http.py

```diff
@@ -0,0 +1,114 @@
+import time
+
+import requests
+import requests.adapters
+import urllib3.util.retry
+
+from ..config import config
+from . import logger
+
+DEFAULT_RETRY_COUNT = 3
+DEFAULT_RETRY_BACKOFF = 1
+
+
+class HTTPSessionWithRetry(requests.Session):
+
+    # make sure to only add exceptions that are raised early in the request. For example, ConnectionError can be raised
+    # during the handling of a request, and therefore should not be retried, as the request might not be idempotent.
+    HTTP_RETRYABLE_EXCEPTIONS = {
+        # ConnectionResetError is raised when the server closes the connection prematurely during TCP handshake.
+        ConnectionResetError: ["Connection reset by peer", "Connection aborted"],
+        # "Connection aborted" and "Connection refused" happen when the server doesn't respond at all.
+        ConnectionError: ["Connection aborted", "Connection refused"],
+        ConnectionRefusedError: ["Connection refused"],
+    }
+
+    def __init__(
+        self,
+        max_retries=DEFAULT_RETRY_COUNT,
+        retry_backoff_factor=DEFAULT_RETRY_BACKOFF,
+        retry_on_exception=True,
+        retry_on_status=True,
+        retry_on_post=False,
+    ):
+        super().__init__()
+
+        self.max_retries = max_retries
+        self.retry_backoff_factor = retry_backoff_factor
+        self.retry_on_exception = retry_on_exception
+
+        if retry_on_status:
+            http_adapter = requests.adapters.HTTPAdapter(
+                max_retries=urllib3.util.retry.Retry(
+                    total=self.max_retries,
+                    backoff_factor=self.retry_backoff_factor,
+                    status_forcelist=[500, 502, 503, 504],
+                    method_whitelist=False
+                    if retry_on_post
+                    else urllib3.util.retry.Retry.DEFAULT_METHOD_WHITELIST,
```

Comment:
Maybe should move to helper func or add comments, choose whatever you think is clearer.